firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteDataBackfill.java

```diff
@@ -0,0 +1,177 @@
+
+  /**
+   * Populates the read time used during Index-Free query processing.
+   *
+   * @return Whether the migration finished.
+   */
+  boolean populateReadTime() {
+    boolean[] done = new boolean[] {false};
+
+    SQLitePersistence.Query watermarkQuery =
+        new SQLitePersistence.Query(
+            db, "SELECT first_document_without_read_time FROM target_globals");
+    watermarkQuery.first(
+        value -> {
+          // The read time migration sets the watermark to NULL when the migration
+          // completes.
+          if (value.isNull(0)) {
+            Logger.debug(LOG_TAG, "No read times to backfill");
+            return;
+          }
+
+          int[] rowCount = new int[] {0};
+          String[] lastMigratedKey = new String[] {null};
+
+          String encodedFirstDocumentWithoutReadTime = value.getString(0);
+          SQLitePersistence.Query query =
+              new SQLitePersistence.Query(
+                      db,
+                      "SELECT path, contents FROM remote_documents "
+                          + "WHERE read_time_seconds IS NULL AND path >= ? ORDER BY path LIMIT ?")
+                  .binding(encodedFirstDocumentWithoutReadTime, BACKFILL_MIGRATION_SIZE);
+
+          query.forEach(
+              cursor -> {
+                ++rowCount[0];
+
+                String encodedCurrentPath = cursor.getString(0);
+                try {
+                  Timestamp readTime = extractReadTime(cursor.getBlob(1));
+                  if (readTime == null) {
+                    Logger.warn(
+                        LOG_TAG,
+                        "Failed to detect document for key %s during read time backfill",
+                        EncodedPath.decodeFieldPath(encodedCurrentPath));
+                    return;
+                  }
+
+                  SQLiteStatement updateStatement =
+                      db.compileStatement(
+                          "UPDATE remote_documents SET read_time_seconds = ? and read_time_nanos = ? WHERE path = ?");
+                  updateStatement.bindLong(1, readTime.getSeconds());
+                  updateStatement.bindLong(2, readTime.getNanos());
+                  updateStatement.bindString(3, encodedCurrentPath);
+                  updateStatement.executeUpdateDelete();
+                  lastMigratedKey[0] = encodedCurrentPath;
+                } catch (InvalidProtocolBufferException e) {
+                  Logger.warn(
+                      LOG_TAG,
+                      "Failed to decode document for key %s during read time backfill",
+                      EncodedPath.decodeFieldPath(encodedCurrentPath));
+                }
+              });
+
+          SQLiteStatement updateStatement =
+              db.compileStatement("UPDATE target_globals SET first_document_without_read_time = ?");
+          if (rowCount[0] == BACKFILL_MIGRATION_SIZE && lastMigratedKey[0] != null) {
+            String nextKey = EncodedPath.prefixSuccessor(lastMigratedKey[0]);
```

Comment:
I'm a little nervous about this prefixSuccessor stuff.  Is there a reason we couldn't just store the last_migrated_key and then do a `>` query instead of `>=` ?

Maybe the issue is differentiating the "not-started" from "finished" state?  Perhaps we could:
* Store "" initially to indicate that we need to migrate the first doc with key > "".
* Store valid keys while migration is in progress.
* Store NULL once the migration is done.